File: go/binarysearch/4_find_median_sorted_arrays.go
package binarysearch

// findMedianSortedArrays 寻找两个正序数组的中位数
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {

	// 偶数
	if total := len(nums1) + len(nums2); total&1 == 0 {
		e1 := getKNums(nums1, nums2, total/2)
		e2 := getKNums(nums1, nums2, total/2+1)
		return float64(e1+e2) / 2
	} else {
		return float64(getKNums(nums1, nums2, total/2))
	}

}

// getKNums 在有序数组nums1/nums2 中获取第k 小的元素(下标从0开始)
func getKNums(nums1, nums2 []int, k int) int {
	// log(m+n) 的复杂度，二分
	// 每次抛弃nums1 nums2 中一般的元素

	for {

		n1, n2 := len(nums1), len(nums2)
		mid1, mid2 := n1>>1, n2>>1

		if n1 == 0 {
			return nums2[k]
		}

		if n2 == 0 {
			return nums1[k]
		}

		// 循环终止的条件, 第0小的元素
		// 每次抛弃一半的元素之后，如果是前半部分的元素，要更新k 的值，会变小
		if k == 0 {
			return min(nums1[0], nums2[0])
		}

		// k在 两个数组的前半部分较小的元素中， 要舍弃后半部分较大的
		// 特别分析，k == mid1 + mid2 的时候，实际上合并有mid1 + 1 + mid2 + 1 个元素
		// 而 第k 小，实际上是 k + 1个元素，因此抛弃元素的时候，mid1/mid2 是可以不要的
		if k <= mid1+mid2 {

			// 说明nums2 的后半部分元素较大，舍弃
			if nums1[mid1] < nums2[mid2] {
				nums2 = nums2[:mid2]
			} else {
				nums1 = nums1[:mid1]
			}

		} else {
			// k 元素在 两个数组中较大的后半部分，要舍弃前半部分较小的
			// 说明 nums1 前半部分较小
			if nums1[mid1] < nums2[mid2] {
				nums1 = nums1[mid1+1:]
				// 更新k 的值, 删除(mid1+1) 个元素，因为mid1 的下标从0开始的
				k -= mid1 + 1
			} else {
				nums2 = nums2[mid2+1:]
				k -= mid2 + 1
			}

		}

	}

}

File: rust/src/binarysearch/_4_find_median_sorted_arrays.rs
/// 寻找两个正序数组的中位数
pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
    let (mut nums1, mut nums2) = (nums1, nums2);

    let l = nums1.len() + nums2.len();
    // 偶数
    if l & 1 == 0 {
        let e1 = get_k_nums(&mut nums1, &mut nums2, l / 2);
        let e2 = get_k_nums(&mut nums1, &mut nums2, l / 2 - 1);
        return (e1 as f64 + e2 as f64) / 2.0;
    } else {
        return get_k_nums(&mut nums1, &mut nums2, l / 2) as f64;
    }
}

/// 返回两个有序数组合并后第k小的元素， 计数从0开始
fn get_k_nums(mut nums1: &[i32], mut nums2: &[i32], mut k: usize) -> i32 {
    // let mut nums1 = nums1.as_slice();
    // let mut nums2 = nums2.as_slice();

    loop {
        let (l1, l2) = (nums1.len(), nums2.len());
        let (mid1, mid2) = (l1 >> 1, l2 >> 1);

        if l1 == 0 {
            return nums2[k];
        }

        if l2 == 0 {
            return nums1[k];
        }

        if k == 0 {
            return nums1[0].min(nums2[0]);
        }

        // k 在两个数组的前半部分元素中
        if k <= mid1 + mid2 {
            // 抛弃后面的元素, 保留前面的元素
            if nums1[mid1] < nums2[mid2] {
                nums2 = &nums2[..mid2];
            } else {
                nums1 = &nums1[..mid1];
            }
        } else {
            // k 在后半部分元素
            // 抛弃前面的元素
            if nums1[mid1] < nums2[mid2] {
                nums1 = &nums1[mid1 + 1..];
                // update k ,因为前面少了(mid1+1)元素
                k -= mid1 + 1;
            } else {
                nums2 = &nums2[mid2 + 1..];
                k -= mid2 + 1;
            }
        }
    }
}
